File: patterns/Creational/Prototype/Java/RealWorldExamples/Document.java
package patterns.Creational.Prototype.Java.RealWorldExamples;

/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Document Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to create new instances by cloning 
| existing objects, which can avoid costly initialization steps.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational.Prototype.RealWorldExamples
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-java
|--------------------------------------------------------------------------
*/

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

// Prototype Interface
interface Prototype extends Cloneable {
    Prototype clone();
}

// Concrete Prototype: Document
class Document implements Prototype {
    private String title;
    private String content;
    private Date createdAt;
    private Author author;

    public Document(String title, String content, Author author) {
        this.title = title;
        this.content = content;
        this.author = author;
        this.createdAt = new Date();
        this.author.addDocument(this);
    }

    @Override
    public Document clone() {
        try {
            Document clone = (Document) super.clone();
            clone.title = "Copy of " + this.title;
            clone.createdAt = new Date();
            this.author.addDocument(clone);
            return clone;
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
    }

    public String getDetails() {
        return "Title: " + title + "\nContent: " + content + "\nAuthor: " + author.getName() + "\nCreatedAt: " + createdAt.toString();
    }
}

// The Author class, which has a collection of documents.
class Author {
    private String name;
    private List<Document> documents = new ArrayList<>();

    public Author(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void addDocument(Document document) {
        this.documents.add(document);
    }

    public List<Document> getDocuments() {
        return documents;
    }
}

// Client code that demonstrates the Prototype Design Pattern.
public class Main {
    public static void main(String[] args) {
        // Create an author and a document.
        Author author = new Author("Jane Doe");
        Document originalDocument = new Document("Design Patterns", "Learning the Prototype Pattern.", author);

        // Clone the original document.
        Document clonedDocument = originalDocument.clone();

        // Display details of both documents.
        System.out.println("Original Document:");
        System.out.println(originalDocument.getDetails());

        System.out.println("\nCloned Document:");
        System.out.println(clonedDocument.getDetails());
    }
}

File: patterns/Creational/Prototype/JavaScript/RealWorldExamples/Document.js
/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Document Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to create new instances by cloning 
| existing objects, which can avoid costly initialization steps.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype/RealWorldExamples
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-js
|--------------------------------------------------------------------------
*/

/**
 * Prototype Interface.
 * Defines the `clone` method that all prototypes must implement.
 */
class Prototype {
    clone() {
        throw new Error("Method 'clone()' must be implemented.");
    }
}

/**
 * Concrete Prototype: Document
 */
class Document extends Prototype {
    constructor(title, content, author) {
        super();
        this.title = title;
        this.content = content;
        this.author = author;
        this.createdAt = new Date();
        this.author.addDocument(this);
    }

    /**
     * Control what data to copy when cloning.
     * Example:
     * - The cloned document will have "Copy of ..." title.
     * - The author reference remains the same.
     * - The creation date is updated to the current date.
     */
    clone() {
        const clone = Object.create(Object.getPrototypeOf(this));
        clone.title = "Copy of " + this.title;
        clone.content = this.content;
        clone.author = this.author;
        clone.createdAt = new Date();
        // Add the cloned document to the author's list of documents.
        this.author.addDocument(clone);
        return clone;
    }

    getDetails() {
        return {
            Title: this.title,
            Content: this.content,
            Author: this.author.getName(),
            CreatedAt: this.createdAt.toISOString().slice(0, 19).replace('T', ' ')
        };
    }
}

/**
 * The Author class, which has a collection of documents.
 */
class Author {
    constructor(name) {
        this.name = name;
        this.documents = [];
    }

    getName() {
        return this.name;
    }

    addDocument(document) {
        this.documents.push(document);
    }

    getDocuments() {
        return this.documents;
    }
}

/**
 * Client code that demonstrates the Prototype Design Pattern.
 */
function clientCode() {
    // Create an author and a document.
    const author = new Author("Jane Doe");
    const originalDocument = new Document("Design Patterns", "Learning the Prototype Pattern.", author);

    // Clone the original document.
    const clonedDocument = originalDocument.clone();

    // Display details of both documents.
    console.log("Original Document:");
    console.log(originalDocument.getDetails());

    console.log("\nCloned Document:");
    console.log(clonedDocument.getDetails());
}

clientCode();
